exonum-java-binding-common/src/test/java/com/exonum/binding/common/proofs/list/ListProofHashCodeCalculatorTest.java

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import static com.exonum.binding.common.hash.Funnels.hashCodeFunnel;
+import static com.google.common.collect.ImmutableMap.of;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+import com.exonum.binding.common.hash.HashCode;
+import com.exonum.binding.common.hash.Hashing;
+import com.exonum.binding.common.hash.PrimitiveSink;
+import com.exonum.binding.common.serialization.StandardSerializers;
+import java.nio.charset.Charset;
+import org.junit.jupiter.api.Test;
+
+class ListProofHashCodeCalculatorTest {
+
+  private static final String V1 = "v1";
+  private static final String V2 = "v2";
+  private static final String V3 = "v3";
+  private static final String V4 = "v4";
+
+  private static final HashCode H1 = HashCode.fromString("a1");
+  private static final HashCode H2 = HashCode.fromString("a2");
+
+  private ListProofRootHashCalculator<String> calculator;
+
+  @Test
+  void visit_SingletonListProof() {
+    ListProof root = leafOf(V1);
+
+    calculator = createListProofCalculator();
+    root.accept(calculator);
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1)));
+    assertEquals(getNodeHashCode(V1), calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof2elements() {
+    ListProofElement left = leafOf(V1);
+    ListProofElement right = leafOf(V2);
+    ListProofBranch root = new ListProofBranch(left, right);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    //calculate expected root hash
+    HashCode expectedRootHash = getBranchHashCode(getNodeHashCode(V1), getNodeHashCode(V2));
+
+    assertThat(calculator.getElements(), equalTo(of(0L, V1,
+        1L, V2)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof4elements() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            leafOf(V2)
+        ),
+        new ListProofBranch(
+            leafOf(V3),
+            leafOf(V4)
+        )
+    );
+
+    //calculate expected root hash
+    HashCode leftBranchHash = getBranchHashCode(getNodeHashCode(V1), getNodeHashCode(V2));
+    HashCode rightBranchHash = getBranchHashCode(getNodeHashCode(V3), getNodeHashCode(V4));
+    HashCode expectedRootHash = getBranchHashCode(leftBranchHash, rightBranchHash);
+
+    calculator = createListProofCalculator();
+    calculator.visit(root);
+
+    assertThat(calculator.getElements(),
+        equalTo(of(0L, V1,
+            1L, V2,
+            2L, V3,
+            3L, V4)));
+    assertEquals(expectedRootHash, calculator.getCalculatedRootHash());
+  }
+
+  @Test
+  void visit_FullProof2elementsHashMismatch() {
```

Comment:
What does it test?

Reply:
Just a case with hash calculation mismatch.

Reply:
ListProofCalculator is no longer responsible for comparing hashes. It does the same as in a test above, but just produces another root hash, because it uses another hash function. One can say we test a hash function here to be different from the one used in getBranchHashCode

Reply:
:warning: